profiler/src/ProfilerEngine/Datadog.Profiler.Native/ContentionProvider.cpp

```diff
@@ -112,13 +112,77 @@ void ContentionProvider::OnContention(std::chrono::nanoseconds contentionDuratio
     }
 
     auto [blockingThreadId, blockingThreadName] = currentThreadInfo->SetBlockingThread(0, WStr(""));
-    AddContentionSample(0ns, -1, contentionDuration, blockingThreadId, std::move(blockingThreadName), _emptyStack);
+    AddContentionSample(0ns, -1, ContentionType::Lock, contentionDuration, blockingThreadId, std::move(blockingThreadName), _emptyStack);
 }
 
-void ContentionProvider::AddContentionSample(std::chrono::nanoseconds timestamp, uint32_t threadId, std::chrono::nanoseconds contentionDuration, uint64_t blockingThreadId, shared::WSTRING blockingThreadName, const std::vector<uintptr_t>& stack)
+void ContentionProvider::OnWaitStart(std::chrono::nanoseconds timestamp, uintptr_t associatedObjectId)
+{
+    auto currentThreadInfo = ManagedThreadInfo::CurrentThreadInfo;
+    if (currentThreadInfo == nullptr)
+    {
+        return;
+    }
+
+    // TOO BAD: try to get the type of associatedObjectId to make the difference between Monitor/lock, AutoResetEvent, ManualResetEvent, Mutex and Semaphore
+    // the following code does not work because GetClassFromObject returns CORPROF_E_UNSUPPORTED_CALL_SEQUENCE here
+    //  ClassID classId = 0;
+    //  HRESULT hr = _pCorProfilerInfo->GetClassFromObject(static_cast<ObjectID>(associatedObjectId), &classId);
+    //  if (SUCCEEDED(hr))
+    //  {
+    //      std::string typeName;
+    //      if (_pFrameStore->GetTypeName(classId, typeName))
+    //      {
+    //          std::cout << "WaitStart: " << typeName << std::endl;
+    //      }
+    //  }
+
+    currentThreadInfo->SetWaitStart(timestamp);
+
+    // we can't even compare the ClassID of the associatedObjectId to figure out the type of wait...
+    currentThreadInfo->SetContentionType(ContentionType::Wait);
+}
+
+void ContentionProvider::OnWaitStop(std::chrono::nanoseconds timestamp)
```

Comment:
I have something in my mind:
I wonder if mixing wait events with contention will break the upscaling 🤔 
More to that, in this function you skip some events which means that the upscaling can break.

Reply:
Good point!
Let's have 2 samplers to avoid missing long locks due to long wait by Manual/AutoResetEvents